src/control/lib/netdetect/netdetect.go

```diff
@@ -177,3 +557,276 @@ func GetDeviceNames() ([]string, error) {
+	// GetAffinityForDevice() is used to provide this conversion.
+	for ; fi != nil; fi = fi.next {
+		if fi.domain_attr == nil || fi.domain_attr.name == nil {
+			continue
+		}
+		if fi.fabric_attr == nil || fi.fabric_attr.prov_name == nil {
+			continue
+		}
+		deviceScanCfg.targetDevice = C.GoString(fi.domain_attr.name)
+		deviceAffinity, err := GetAffinityForDevice(deviceScanCfg)
+		if err != nil {
+			continue
+		}
+		if deviceAffinity.DeviceName == device {
+			log.Debugf("Device %s supports provider: %s", device, provider)
+			if deviceAffinity.NUMANode != numaNode {
+				return errors.Errorf("The NUMA node for device %s does not match the provided value %d. " +
+					"Remove the pinned_numa_node value from daos_server.yml then execute 'daos_server network scan' " +
+					"to see the valid NUMA node associated with the network device", device, numaNode)
+			}
+			log.Debugf("The NUMA node for device %s matches the provided value %d.  Network configuration is valid.", device, numaNode)
+			return nil
+		}
+	}
+	return errors.Errorf("Device %s does not support provider: %s", device, provider)
+}
+
+// ScanFabric examines libfabric data to find the network devices that support the given fabric provider.
+func ScanFabric(provider string) ([]FabricScan, error) {
+	var ScanResults []FabricScan
+	var fi *C.struct_fi_info
+	var hints *C.struct_fi_info
+	var devCount int
+
+	resultsMap := make(map[string]struct{})
+
+	hints = C.fi_allocinfo()
+	if hints == nil {
+		return ScanResults, errors.New("unable to initialize lib fabric - failed to allocinfo")
+	}
+	defer C.fi_freeinfo(hints)
+
+	// If a provider was given, then set the libfabric search hint to match the provider
+	if (provider != "") {
+		log.Debugf("Input provider string: %s", provider)
+		libfabricProviderList, err := ConvertMercuryToLibFabric(provider)
+		if err != nil {
+			return ScanResults, err
+		}
+		log.Debugf("Final libFabricProviderList is %s", libfabricProviderList)
+		hints.fabric_attr.prov_name = C.strdup(C.CString(libfabricProviderList))
+	}
+
+	// Initialize libfabric and perform the scan
+	C.fi_getinfo(C.uint(libFabricMajorVersion << 16 | libFabricMinorVersion), nil, nil, 0, hints, &fi)
+	if (fi == nil) {
+		return ScanResults, errors.New("libfabric found no records matching the specified provider")
+	}
+	defer C.fi_freeinfo(fi)
+
+	// We have some data from libfabric scan.  Let's initialize hwloc and get the device scan started
+	deviceScanCfg, err := initDeviceScan()
+	if err != nil {
+		return ScanResults, err
+	}
+	defer cleanUp(deviceScanCfg.topology)
+
+	log.Debugf("initDeviceScan completed.  Depth %d, numObj %d, systemDeviceNames %v, hwlocDeviceNames %v",
+		deviceScanCfg.depth, deviceScanCfg.numObj, deviceScanCfg.systemDeviceNames, deviceScanCfg.hwlocDeviceNames)
+
+	for ; fi != nil; fi = fi.next {
+		if fi.domain_attr == nil || fi.domain_attr.name == nil {
```

Comment:
as above, could this check be combined with next one

Reply:
Done.